### PR TITLE
Join with pois by dist and datetime

### PR DESCRIPTION
**Pull request by pedroccufc**
## Types of changes

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [ ] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to change)
-   [ ] I have read the **CONTRIBUTING** document.
-   [x] My code follows the code style of this project (see `.code-style.md`).
-   [ ] All new and existing tests passed (see `.testing.md`).
-   [ ] I have added tests to cover my changes.
-   [ ] My change requires a change to the documentation.
-   [ ] I have updated the documentation accordingly.

## Description

-   What is the current behavior? (You can also link to an open issue here)
    

-   What is the new behavior (if this is a feature change)?

### CODE REVIEW

File: pymove/utils/integration.py [hunk 1]
```diff
@@ -697,6 +697,121 @@ def join_with_poi_datetime_optimizer(
...
     except Exception as e:
         raise e
+        
+        
+def join_with_pois_by_dist_and_datetime(
+    df_, 
+    df_pois, 
+    label_date=DATETIME, 
+    label_event_id=EVENT_ID, 
+    label_event_type=EVENT_TYPE,
```
Comment (regispires) [inline pymove/utils/integration.py:707]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/InsightLab/PyMove/pullRequest?prid=5857559)

File: pymove/utils/integration.py [hunk 2]
```diff
@@ -697,6 +697,121 @@ def join_with_poi_datetime_optimizer(
 
     except Exception as e:
         raise e
+        
+        
+def join_with_pois_by_dist_and_datetime(
+    df_, 
+    df_pois,
```
Comment (regispires) [inline pymove/utils/integration.py:704]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/InsightLab/PyMove/pullRequest?prid=5857559)

File: pymove/utils/integration.py [hunk 3]
```diff
@@ -697,6 +697,121 @@ def join_with_poi_datetime_optimizer(
 
     except Exception as e:
         raise e
+        
+        
+def join_with_pois_by_dist_and_datetime(
+    df_, 
+    df_pois, 
+    label_date=DATETIME,
```
Comment (regispires) [inline pymove/utils/integration.py:705]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/InsightLab/PyMove/pullRequest?prid=5857559)

File: pymove/utils/integration.py [hunk 4]
```diff
@@ -697,6 +697,121 @@ def join_with_poi_datetime_optimizer(
...
+            # set min and max of coordinates by radius
+            latmin, lonmin, latmax, lonmax = filters.get_bbox_by_radius((df_.at[idx, LATITUDE], df_.at[idx, LONGITUDE]), radius)
+            
+            # filter event by radius
+            df_filtered = filters.by_bbox(df_pois, (latmin, lonmin, latmax, lonmax))
+            
+            # filter event by datetime
+            filters.by_datetime(
+                df_filtered, 
+                start_datetime=window_start[idx],
```
Comment (regispires) [inline pymove/utils/integration.py:764]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/InsightLab/PyMove/pullRequest?prid=5857559)

File: pymove/utils/integration.py [hunk 5]
```diff
@@ -697,6 +697,121 @@ def join_with_poi_datetime_optimizer(
...
+                df_filtered.reset_index(drop=True, inplace=True)
+
+                # create lat and lon array to operation
+                lat_user = np.full(size_filter, df_.at[idx, LATITUDE], dtype=np.float64)
+                lon_user = np.full(size_filter, df_.at[idx, LONGITUDE], dtype=np.float64)
+
+                # calculate of distances between points
+                distances = haversine(
+                    lat_user, 
+                    lon_user,
```
Comment (regispires) [inline pymove/utils/integration.py:783]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/InsightLab/PyMove/pullRequest?prid=5857559)

File: pymove/utils/integration.py [hunk 6]
```diff
@@ -697,6 +697,121 @@ def join_with_poi_datetime_optimizer(
...
+
+            # set min and max of coordinates by radius
+            latmin, lonmin, latmax, lonmax = filters.get_bbox_by_radius((df_.at[idx, LATITUDE], df_.at[idx, LONGITUDE]), radius)
+            
+            # filter event by radius
+            df_filtered = filters.by_bbox(df_pois, (latmin, lonmin, latmax, lonmax))
+            
+            # filter event by datetime
+            filters.by_datetime(
+                df_filtered,
```
Comment (regispires) [inline pymove/utils/integration.py:763]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/InsightLab/PyMove/pullRequest?prid=5857559)

File: pymove/utils/integration.py [hunk 7]
```diff
@@ -697,6 +697,121 @@ def join_with_poi_datetime_optimizer(
...
+                # reseting index of data frame
+                df_filtered.reset_index(drop=True, inplace=True)
+
+                # create lat and lon array to operation
+                lat_user = np.full(size_filter, df_.at[idx, LATITUDE], dtype=np.float64)
+                lon_user = np.full(size_filter, df_.at[idx, LONGITUDE], dtype=np.float64)
+
+                # calculate of distances between points
+                distances = haversine(
+                    lat_user,
```
Comment (regispires) [inline pymove/utils/integration.py:782]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/InsightLab/PyMove/pullRequest?prid=5857559)

File: pymove/utils/integration.py [hunk 8]
```diff
@@ -697,6 +697,121 @@ def join_with_poi_datetime_optimizer(
...
+
+                # create lat and lon array to operation
+                lat_user = np.full(size_filter, df_.at[idx, LATITUDE], dtype=np.float64)
+                lon_user = np.full(size_filter, df_.at[idx, LONGITUDE], dtype=np.float64)
+
+                # calculate of distances between points
+                distances = haversine(
+                    lat_user, 
+                    lon_user, 
+                    df_filtered[LATITUDE].to_numpy(),
```
Comment (regispires) [inline pymove/utils/integration.py:784]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/InsightLab/PyMove/pullRequest?prid=5857559)

File: pymove/utils/integration.py [hunk 9]
```diff
@@ -697,6 +697,121 @@ def join_with_poi_datetime_optimizer(
 
     except Exception as e:
         raise e
+        
+        
+def join_with_pois_by_dist_and_datetime(
+    df_, 
+    df_pois, 
+    label_date=DATETIME, 
+    label_event_id=EVENT_ID,
```
Comment (regispires) [inline pymove/utils/integration.py:706]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/InsightLab/PyMove/pullRequest?prid=5857559)

File: pymove/utils/integration.py [hunk 10]
```diff
@@ -697,6 +697,121 @@ def join_with_poi_datetime_optimizer(
...
+    df_pois, 
+    label_date=DATETIME, 
+    label_event_id=EVENT_ID, 
+    label_event_type=EVENT_TYPE, 
+    time_window=3600,
+    radius=1000
+):
+    """
+    It performs the integration between trajectories and points of interest,
+    generating new columns referring to the category of the point of interest,
```
Comment (regispires) [inline pymove/utils/integration.py:713]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/InsightLab/PyMove/pullRequest?prid=5857559)

File: pymove/utils/integration.py [hunk 11]
```diff
@@ -697,6 +697,121 @@ def join_with_poi_datetime_optimizer(
 
     except Exception as e:
         raise e
+        
+        
+def join_with_pois_by_dist_and_datetime(
+    df_,
```
Comment (regispires) [inline pymove/utils/integration.py:703]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/InsightLab/PyMove/pullRequest?prid=5857559)

File: pymove/utils/integration.py [hunk 12]
```diff
@@ -697,6 +697,121 @@ def join_with_poi_datetime_optimizer(
 
     except Exception as e:
         raise e
+
```
Comment (regispires) [inline pymove/utils/integration.py:700]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/InsightLab/PyMove/pullRequest?prid=5857559)

File: pymove/utils/integration.py [hunk 13]
```diff
@@ -699,6 +745,110 @@ def join_with_poi_datetime_optimizer(
...
+    try:
+        print('Integration with Events...')
+
+        values = _reset_set_window_and_creates_event_id_type_all(
+            df_, df_pois, label_date, time_window
+        )
+
+        window_start, window_end, current_distances, event_id, event_type = values
+
+        for idx, row in progress_bar(df_.iterrows(), total=df_.shape[0]):
```
Comment (regispires) [inline pymove/utils/integration.py:796]: Codacy found an issue: [Unused variable 'row'](https://app.codacy.com/gh/InsightLab/PyMove/pullRequest?prid=5857559)

File: pymove/preprocessing/filters.py [hunk 14]
```diff
@@ -14,6 +14,43 @@
...
+
+    radius: Float, optional (1000 by default)
+
+    Returns
+    -------
+    array
+        coordinates min and max
+
+    references:
+        https://mathmesquita.me/2017/01/16/filtrando-localizacao-em-um-raio.html
```
Comment (flych3r) [inline pymove/preprocessing/filters.py:32]: Segue o mesmo esquema de Parameters e Returns
```python
"""
...
References
---------------
    https://mathmesquita.me/2017/01/16/filtrando-localizacao-em-um-raio.html
"""
```
Não sei pq a visualização ta mostrano assim, mas o tamanho dos `-` é o mesmo de `References`

File: pymove/preprocessing/filters.py [hunk 15]
```diff
@@ -14,6 +14,43 @@
 )
 
 
+def get_bbox_by_radius(coordinates, radius=1000):
+    """
```
Comment (flych3r) [inline pymove/preprocessing/filters.py:18]: O início da docstring deve ter um resumo do objetivo da função.

File: pymove/preprocessing/filters.py [hunk 16]
```diff
@@ -14,6 +14,43 @@
...
+    ----------
+    coords: tupla (lat, lon)
+        The coordinates of point
+
+    radius: Float, optional (1000 by default)
+
+    Returns
+    -------
+    array
+        coordinates min and max
```
Comment (flych3r) [inline pymove/preprocessing/filters.py:29]: Enrriquecer a descrição do retorno: tipo: coordenadas min e max da bbox

File: pymove/utils/integration.py [hunk 17]
```diff
@@ -259,6 +259,52 @@ def _reset_set_window__and_creates_event_id_type(
...
+
+    label_date : String
+        Label of df_ referring to the datetime.
+
+    time_window : Int
+        Number of seconds of the time window.
+
+    Returns
+    -------
+    window_starts, window_ends, current_distances, event_id, event_type
```
Comment (flych3r) [inline pymove/utils/integration.py:286]: Descrever os tipos das variáveis com a descrição, como em `get_bbox_by_radius`

File: pymove/utils/integration.py [hunk 18]
```diff
@@ -699,6 +745,110 @@ def join_with_poi_datetime_optimizer(
...
+    label_event_type=EVENT_TYPE,
+    time_window=3600,
+    radius=1000,
+    inplace=False
+):
+    """
+    It performs the integration between trajectories and points of interest,
+    generating new columns referring to the category of the point of interest,
+    the distance between the location of the vehiculo and location of the poi
+    based on the distance and on time of each point of the trajectories.
```
Comment (flych3r) [inline pymove/utils/integration.py:762]: Descrever melhor a função, para deixar claro que a união é feita referente ao tempo de ambos o ponto e o poi.
Trocar `vehiculo` por `user`.

File: pymove/utils/integration.py [hunk 19]
```diff
@@ -699,6 +745,110 @@ def join_with_poi_datetime_optimizer(
...
+        Label of df_events referring to the type of the event.
+
+    time_window : Float, optional(3600 by default)
+        tolerable length of time range for assigning the event's
+        point of interest to the trajectory point.
+
+    radius: Float, optional (1000 by default)
+    """
+    try:
+        print('Integration with Events...')
```
Comment (flych3r) [inline pymove/utils/integration.py:788]: Verificar se o dataframe de poi possui a coluna de dataframe, se não possuir lançar uma exceção, pois ela é necessária pra função.
Pode ser tipo assim:
```python
if label_date not in df_pois:
    raise KeyError('POI's dataframe must contain a %s column' % label_date)
```

File: pymove/__init__.py [hunk 20]
```diff
@@ -26,6 +26,7 @@
     mem,
     trajectories,
     transformations,
+    visual,
 )
 from .utils.trajectories import read_csv
-from .visualization import visualization
+from .visualization import folium, matplotlib
```
Comment (regispires) [inline pymove/__init__.py:32]: Codacy found an issue: ['.visualization.folium' imported but unused (F401)](https://app.codacy.com/gh/InsightLab/PyMove/pullRequest?prid=5857559)

File: pymove/visualization/matplotlib.py [hunk 21]
```diff
@@ -0,0 +1,213 @@
...
+def show_object_id_by_date(
+    move_data,
+    create_features=True,
+    kind=None,
+    figsize=(21, 9),
+    return_fig=True,
+    save_fig=True,
+    name='shot_points_by_date.png',
+):
+    """
```
Comment (regispires) [inline pymove/visualization/matplotlib.py:24]: Codacy found an issue: [Multi-line docstring summary should start at the first line (D212)](https://app.codacy.com/gh/InsightLab/PyMove/pullRequest?prid=5857559)

File: pymove/__init__.py (hunk 20 above)
Comment (regispires) [inline pymove/__init__.py:32]: Codacy found an issue: ['.visualization.folium' imported but unused (F401)](https://app.codacy.com/gh/InsightLab/PyMove/pullRequest?prid=5857559)

File: pymove/visualization/matplotlib.py (hunk 21 above)
Comment (regispires) [inline pymove/visualization/matplotlib.py:24]: Codacy found an issue: [Multi-line docstring summary should start at the first line (D212)](https://app.codacy.com/gh/InsightLab/PyMove/pullRequest?prid=5857559)

File: pymove/__init__.py (hunk 20 above)
Comment (regispires) [inline pymove/__init__.py:32]: Codacy found an issue: ['.visualization.folium' imported but unused (F401)](https://app.codacy.com/gh/InsightLab/PyMove/pullRequest?prid=5857559)

File: pymove/visualization/matplotlib.py (hunk 21 above)
Comment (regispires) [inline pymove/visualization/matplotlib.py:24]: Codacy found an issue: [Multi-line docstring summary should start at the first line (D212)](https://app.codacy.com/gh/InsightLab/PyMove/pullRequest?prid=5857559)

File: pymove/__init__.py (hunk 20 above)
Comment (regispires) [inline pymove/__init__.py:32]: Codacy found an issue: ['.visualization.folium' imported but unused (F401)](https://app.codacy.com/gh/InsightLab/PyMove/pullRequest?prid=5857559)

File: pymove/visualization/matplotlib.py (hunk 21 above)
Comment (regispires) [inline pymove/visualization/matplotlib.py:24]: Codacy found an issue: [Multi-line docstring summary should start at the first line (D212)](https://app.codacy.com/gh/InsightLab/PyMove/pullRequest?prid=5857559)

File: pymove/tests/test_utils_integration.py [hunk 22]
```diff
@@ -598,6 +598,8 @@ def test_join_with_poi_datetime():
     )
 
     integration.join_with_poi_datetime(move_df, pois, time_window=45000)
+    print(move_df.columns)
+    print(expected.columns)
```
Comment (flych3r) [inline pymove/tests/test_utils_integration.py]: Remover debub

File: pymove/__init__.py (hunk 20 above)
Comment (regispires) [inline pymove/__init__.py:32]: Codacy found an issue: ['.visualization.folium' imported but unused (F401)](https://app.codacy.com/gh/InsightLab/PyMove/pullRequest?prid=5857559)

File: pymove/visualization/matplotlib.py (hunk 21 above)
Comment (regispires) [inline pymove/visualization/matplotlib.py:24]: Codacy found an issue: [Multi-line docstring summary should start at the first line (D212)](https://app.codacy.com/gh/InsightLab/PyMove/pullRequest?prid=5857559)

File: pymove/__init__.py (hunk 20 above)
Comment (regispires) [inline pymove/__init__.py:32]: Codacy found an issue: ['.visualization.folium' imported but unused (F401)](https://app.codacy.com/gh/InsightLab/PyMove/pullRequest?prid=5857559)

File: pymove/visualization/matplotlib.py (hunk 21 above)
Comment (regispires) [inline pymove/visualization/matplotlib.py:24]: Codacy found an issue: [Multi-line docstring summary should start at the first line (D212)](https://app.codacy.com/gh/InsightLab/PyMove/pullRequest?prid=5857559)

File: pymove/utils/trajectories.py [hunk 23]
```diff
@@ -255,19 +275,80 @@ def fill_list_with_new_values(original_list, new_list_values):
...
+        The input trajectory data.
+
+    label_conversion : Object
+        Label of df_ referring to the column for conversion.
+    """
+    try:
+
+        if label_conversion not in df_:
+            raise KeyError(
+                "POI's dataframe must contain a %s column" % label_conversion
```
Comment (flych3r) [inline pymove/utils/trajectories.py:324]: Pode ser qualquer dataframe, não apenas POI.

File: pymove/utils/trajectories.py [hunk 24]
```diff
@@ -255,19 +275,80 @@ def fill_list_with_new_values(original_list, new_list_values):
...
+        object converted to a list
+    """
+
+    if object_ is None:
+        return object_
+
+    try:
+        conv = np.array(object_[1:-1].split(', '))
+
+        if conv[0].isdigit():
```
Comment (flych3r) [inline pymove/utils/trajectories.py:299]: E se a lista for de floats?

File: pymove/utils/math.py [hunk 25]
```diff
@@ -1,6 +1,27 @@
 import math
 
 
+def is_number(value):
+    """
```
Comment (regispires) [inline pymove/utils/math.py:5]: Codacy found an issue: [First line should end with a period, question mark, or exclamation point (not 'n') (D415)](https://app.codacy.com/gh/InsightLab/PyMove/pullRequest?prid=5857559)

File: pymove/__init__.py (hunk 20 above)
Comment (regispires) [inline pymove/__init__.py:32]: Codacy found an issue: ['.visualization.folium' imported but unused (F401)](https://app.codacy.com/gh/InsightLab/PyMove/pullRequest?prid=5857559)

File: pymove/visualization/matplotlib.py (hunk 21 above)
Comment (regispires) [inline pymove/visualization/matplotlib.py:24]: Codacy found an issue: [Multi-line docstring summary should start at the first line (D212)](https://app.codacy.com/gh/InsightLab/PyMove/pullRequest?prid=5857559)

**Pull request merged**